diff -r b36eeab1df8b editor/libeditor/base/nsEditPropertyAtomList.h
--- a/editor/libeditor/base/nsEditPropertyAtomList.h	Sat Sep 25 04:02:00 2010 -0400
+++ b/editor/libeditor/base/nsEditPropertyAtomList.h	Mon Oct 18 10:51:27 2010 +0200
@@ -145,6 +145,7 @@
 EDITOR_ATOM(img, "img")
 EDITOR_ATOM(input, "input")
 EDITOR_ATOM(kbd, "kbd")
+EDITOR_ATOM(keygen, "keygen")
 EDITOR_ATOM(label, "label")
 EDITOR_ATOM(legend, "legend")
 EDITOR_ATOM(li, "li")
diff -r b36eeab1df8b editor/libeditor/html/nsHTMLEditUtils.cpp
--- a/editor/libeditor/html/nsHTMLEditUtils.cpp	Sat Sep 25 04:02:00 2010 -0400
+++ b/editor/libeditor/html/nsHTMLEditUtils.cpp	Mon Oct 18 10:51:27 2010 +0200
@@ -432,6 +432,7 @@
       || (nodeAtom == nsEditProperty::select)
       || (nodeAtom == nsEditProperty::button)
       || (nodeAtom == nsEditProperty::output)
+      || (nodeAtom == nsEditProperty::keygen)
       || (nodeAtom == nsEditProperty::input);
 }
 
@@ -651,7 +652,7 @@
   ELEM(isindex, PR_FALSE, PR_FALSE, GROUP_BLOCK | GROUP_HEAD_CONTENT,
        GROUP_NONE),
   ELEM(kbd, PR_TRUE, PR_TRUE, GROUP_PHRASE, GROUP_INLINE_ELEMENT),
-  ELEM(keygen, PR_FALSE, PR_FALSE, GROUP_NONE, GROUP_NONE),
+  ELEM(keygen, PR_FALSE, PR_FALSE, GROUP_FORMCONTROL, GROUP_NONE),
   ELEM(label, PR_TRUE, PR_FALSE, GROUP_FORMCONTROL, GROUP_INLINE_ELEMENT),
   ELEM(legend, PR_TRUE, PR_TRUE, GROUP_NONE, GROUP_INLINE_ELEMENT),
   ELEM(li, PR_TRUE, PR_FALSE, GROUP_LI, GROUP_FLOW_ELEMENT),
@@ -775,7 +776,7 @@
   }
 
   // Deprecated elements.
-  if (aChild == eHTMLTag_bgsound || aChild == eHTMLTag_keygen) {
+  if (aChild == eHTMLTag_bgsound) {
     return PR_FALSE;
   }
 
